store/tikv/coprocessor.go

```diff
@@ -832,6 +852,33 @@ func (it *copIterator) Close() error {
 	return nil
 }
 
+type rateLimit struct {
+	token chan struct{}
+}
+
+func newRateLimit(n int) *rateLimit {
+	return &rateLimit{
+		token: make(chan struct{}, n),
+	}
+}
+
+func (r *rateLimit) getToken(done <-chan struct{}) (exit bool) {
+	select {
+	case <-done:
+		return true
+	case r.token <- struct{}{}:
+		return false
+	}
+}
+
+func (r *rateLimit) putToken() {
+	select {
+	case <-r.token:
+	default:
+		panic("put a redundant token")
```

Comment:
It's better to return an error instead of just panic.

Reply:
It's a `not reachable` path and we should panic here.
An example: https://github.com/golang/go/blob/master/src/bufio/bufio.go#L395-L396

Reply:
It's unreachable only when programmer writes code correctly. But everyone makes mistake. It's better to return an error to make our user notice that we have made a mistake. And that mistake should not impact the long running database service.

Reply:
There are several cons using error but not panic:
1. `rateLimit.putToken` is dirty with an unmeaning error return value.
2. The user of `rateLimit.putToken` has to do a redundant error check.
3. Make a bug hard and late to reveal.

Besides, we should consider using `recover` to handle mistakes made by programmers.

Reply:
You can see many cases with `panic("unreachable")` in go world and every mistake make by programmers may panic and let bugs revealed.

https://github.com/golang/go/search?q=panic%28%22unreachable%22%29&unscoped_q=panic%28%22unreachable%22%29

Reply:
> 1. rateLimit.putToken is dirty with an unmeaning error return value.
> 2. The user of rateLimit.putToken has to do a redundant error check.
> 3. Make a bug hard and late to reveal.

IMHO,
Golang is a tool that the programmers will use, while TiDB is a tool that the customers will use.
These 3 points are all from the perspective of a programmer, but not from the perspective of a customer.
We need to make the behavior be more acceptable to the customers.

Reply:
Agreed with @XuHuaiyu. That's why we changed most of the panic in our codebase to error in the past few PRs.

Reply:
For customers, we should use `defer` to handle this sort of panic.
like https://github.com/pingcap/tidb/blob/master/server/conn.go#L1258-L1276

Reply:
`recover()` on that goroutine doesn't work. If you are going to use the `panic-recover` mechanism, you should `recover` in the gogoutine where the function caller belongs to.

Reply:
Otherwise, in this circumstance, if a token cannot be put into the channel, it may lead to unexpected behavior, I do not think the caller could handle the expectation correctly.

Reply:
> `recover()` on that goroutine doesn't work. If you are going to use the `panic-recover` mechanism, you should `recover` in the gogoutine where the function caller belongs to.

Yes, it's recovered by https://github.com/pingcap/tidb/blob/730e140/server/conn.go#L1258-L1276 with `streamResult` and  https://github.com/pingcap/tidb/blob/730e1408cfbf75ed3c45fd98e3a8bdcf5b15d89a/distsql/select_result.go#L90-L102 with `selectResult`.

Reply:
okay.